File: gisrenderer/gismaplib/src/gismaplib/messages/GetTestTextMsg.cpp
#include "gismaplib/messages/GetTestTextMsg.h"

#include "gismaplib/GmLog.h"
#include "gismaplib/utils/SharedPointers.h"


namespace gmcore
{
  std::string* bigStr = nullptr;

  std::string* getBigStr()
  {
    std::string* src = new std::string("abcdefghi ");  // 10 chars
    std::string* dst = new std::string();

    for (int i = 0; i < 100000; ++i) {
      // 10 * 100 000 = 1 000 000 chars
      *dst += *src;
    }

    return (dst);
  }  // getBigStr


  void Deleter(std::string* ptr) {}


  SharedString getBigTestText(const std::string& testPath)
  {
    if (nullptr == bigStr) {
      bigStr = getBigStr();
      BOOST_LOG_SEV(GmLog::log, debug) << "big str created.";
    }
    std::shared_ptr <std::string> str(bigStr, Deleter);

    return (str);
  }  // getBigTestText


  SharedString getTestText(const std::string& testPath)
  {
    auto str = makeSharedString("test text");
    return (str);
  }


  void GetTestTextMsg::dataWorker(
      boost::shared_ptr <capnp::AnyPointer::Reader> dataPtrQ,
      boost::shared_ptr <GmMsg::Message::Builder>   msgPtrR)
  {
    // get the query data
    GmMsg::GetTestTextQ::Reader dataQ = dataPtrQ->getAs <GmMsg::GetTestTextQ>();
    // TODO: avoid make copy of cStr()
    std::string                 testPath = dataQ.getTestPath().cStr();
//    BOOST_LOG_SEV(GmLog::log, debug) << "testPath: " << testPath;

    // make the reply data
    // get test text
//    BOOST_LOG_SEV(GmLog::log, debug) << "get test text";

    auto testText = getTestText(testPath);
//    auto testText = getBigTestText(testPath);

//    BOOST_LOG_SEV(GmLog::log, debug) << "test text: " << testText->c_str();
//    BOOST_LOG_SEV(GmLog::log, debug) << "c++ byte size: " << testText->size();

    // set the reply data
    if (nullptr != testText) {
      capnp::AnyPointer::Builder   dataPtrR = msgPtrR->initDataPointer();
      GmMsg::GetTestTextR::Builder dataR = dataPtrR.initAs <GmMsg::GetTestTextR>();
      dataR.setTestText(testText->c_str());
    }
    else {
      throw GmCoreErrEx() << GmCoreErrInfo("getTestText() error, nullptr == testText");
    }
  }  // GetTestTextMsg::dataWorker
}  // namespace gmcore

File: gisrenderer/app/src/main/java/com/nextgis/gismaplib/messages/GetTestTextMsg.java
package com.nextgis.gismaplib.messages;

import com.nextgis.gismaplib.capnproto.GmConst;
import com.nextgis.gismaplib.capnproto.GmMsg;
import org.capnproto.AnyPointer;

import java.io.IOException;


public class GetTestTextMsg
    extends GmCoreMsg
{
  protected static final String TAG = GetTestTextMsg.class.getName();

  protected String mTestPath;
  protected String mTestText = null;


  public GetTestTextMsg(String testPath)
  {
    super();
    mMsgType = GmConst.MSG_TYPE_GET_TEST_TEXT;
    mTestPath = testPath;
  }


  @Override
  protected void setDataQ(AnyPointer.Builder dataPtrQ)
  {
    // set the query data
    GmMsg.GetTestTextQ.Builder dataQ = dataPtrQ.initAs(GmMsg.GetTestTextQ.factory);
    dataQ.setTestPath(mTestPath);
  }


  @Override
  protected AnyPointer.Reader msgWorker()
      throws IOException
  {
    // get the reply data
    AnyPointer.Reader dataPtrR = super.msgWorker();
    GmMsg.GetTestTextR.Reader dataR = dataPtrR.getAs(GmMsg.GetTestTextR.factory);
    mTestText = dataR.getTestText().toString();
    return (dataPtrR);
  }


  public String getTestText()
      throws IOException
  {
    if (null == mTestText) {
      msgWorker();
    }
    return (mTestText);
  }
}
